core/templates/pages/signup-page/signup-page.component.spec.ts

```diff
@@ -91,30 +90,202 @@ describe('Signup controller', function() {
     }
   });
 
-  it('should show warning if user has not agreed to terms', function() {
-    ctrl.submitPrerequisitesForm(false, null);
-    expect(mockAlertsService.addWarning).toHaveBeenCalledWith(
-      'I18N_SIGNUP_ERROR_MUST_AGREE_TO_TERMS');
+  describe('when has no user logged', function() {
+    beforeEach(function() {
+      $httpBackend.expectGET('/signuphandler/data').respond({
+        username: '',
+        has_agreed_to_latest_terms: false,
+        can_send_emails: true
+      });
+
+      ctrl.$onInit();
+      $httpBackend.flush();
+    });
+
+    it('should show a loading message until the data is retrieved', function() {
+      ctrl.updateWarningText(ctrl.username);
+      ctrl.submitPrerequisitesForm(true, '', true);
+    });
+
+    it('should show warning if email preferences is null but user can send' +
+      ' emails and reset its value', function() {
+      expect(ctrl.emailPreferencesWarningText).toBe(undefined);
+
+      ctrl.submitPrerequisitesForm(true, '', null);
+
+      expect(ctrl.emailPreferencesWarningText).toBe(
+        'I18N_SIGNUP_FIELD_REQUIRED');
+
+      $httpBackend.verifyNoOutstandingExpectation();
+      $httpBackend.verifyNoOutstandingRequest();
+
+      ctrl.onSelectEmailPreference();
+      expect(ctrl.emailPreferencesWarningText).toBe('');
+    });
+
+    it('should send correct informat', function() {
+      spyOn(UrlService, 'getUrlParams').and.returnValue({
+        return_url: '/expected_url'
+      });
+
+      var isRequestTheExpectOne = function(queryParams) {
+        return decodeURIComponent(queryParams).match(
+          '"can_receive_email_updates":true');
+      };
+
+      $httpBackend.expectPOST('/signuphandler/data', isRequestTheExpectOne)
+        .respond(200);
+      ctrl.submitPrerequisitesForm(true, '', 'yes');
+      $httpBackend.flush();
+
+      expect(mockWindow.location.href).toBe('/expected_url');
```

Comment:
Is there something more that we can assert?

Reply:
Sure, done!